oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/token/JwtAccessTokenOAuth2TokenIssuer.java

```diff
@@ -0,0 +1,144 @@
+
+	private final JwtEncoder jwtEncoder;
+
+	private JwtOAuth2TokenCustomizer customizer;
+	private Iterable<JwtOAuth2TokenValidator> validators;
+
+	public JwtAccessTokenOAuth2TokenIssuer(JwtEncoder jwtEncoder) {
+		this.jwtEncoder = jwtEncoder;
+		this.customizer = (header, claims, request) -> {};
+		this.validators = getDefaultValidators();
+	}
+
+	void setClock(Clock clock) {
+		this.clock = clock;
+	}
+
+	/**
+	 * Override default token validators with the provided collection. Validators are run after all customization applied.
+	 *
+	 * @param validators new validators collection
+	 */
+	public void setValidators(Iterable<JwtOAuth2TokenValidator> validators) {
+		this.validators = validators;
+	}
+
+	/**
+	 * Set new token customizer. Customizer is applied before issuing the token and it is a best extension point to implement
+	 * specific token generation logic.
+	 *
+	 * @param customizer new customizer
+	 */
+	public void setCustomizer(JwtOAuth2TokenCustomizer customizer) {
+		this.customizer = customizer;
+	}
+
+	/**
+	 * Issue a new Jwt access token. The generated token's claim names are compliant with RFC7519.
+	 *
+	 * @see OAuth2AccessToken
+	 * @see Jwt
+	 * @see <a href="https://tools.ietf.org/html/rfc7519#section-4">RFC7519 section 4</a>
+	 *
+	 * @param tokenRequest the request
+	 * @return a new OAuth2Token with metadata
+	 */
+	@Override
+	public OAuth2TokenResult<OAuth2AccessToken> issue(@NonNull OAuth2TokenRequest tokenRequest) {
+		Assert.notNull(tokenRequest.getPrincipalName(), "resourceOwner cannot be null");
+		Assert.notNull(tokenRequest.getRegisteredClient(), "registeredClient cannot be null");
+		Assert.notNull(tokenRequest.getTimeToLive(), "timeToLive cannot be null");
+
+		Instant issuedAt = Instant.now(this.clock);
+		Instant expiresAt = issuedAt.plus(tokenRequest.getTimeToLive());
+
+		JoseHeader.Builder joseHeaderBuilder = JoseHeader.withAlgorithm(SignatureAlgorithm.RS256);
+		Set<String> scopes = tokenRequest.getScopes();
+		JwtClaimsSet.Builder jwtClaimsSetBuilder = JwtClaimsSet.withClaims()
+				.subject(tokenRequest.getPrincipalName())
+				.audience(Collections.singletonList(tokenRequest.getRegisteredClient().getClientId()))
+				.issuedAt(issuedAt)
+				.expiresAt(expiresAt)
+				.notBefore(issuedAt)
+				.claim(OAuth2ParameterNames.SCOPE, scopes);
+
+		this.customizer.customize(joseHeaderBuilder, jwtClaimsSetBuilder, tokenRequest);
+
+		JoseHeader joseHeader = joseHeaderBuilder.build();
+		JwtClaimsSet claimsSet = jwtClaimsSetBuilder.build();
+		this.validators.forEach(jwtOAuth2TokenValidator -> jwtOAuth2TokenValidator.validate(joseHeader, claimsSet, tokenRequest));
+
+		Jwt jwt = jwtEncoder.encode(joseHeaderBuilder.build(), jwtClaimsSetBuilder.build());
+		OAuth2AccessToken token = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER, jwt.getTokenValue(), issuedAt, expiresAt, scopes);
+		OAuth2TokenMetadata metadata = OAuth2TokenMetadata
+				.builder()
+				.metadata(OAuth2TokenMetadata.JWT, jwt)
```

Comment:
Let's change `OAuth2TokenMetadata.JWT` to `CLAIMS` to keep things more generic. The type could simply be a `Map`

Reply:
Changed to `TOKEN`. I'm not sure there is much value to store `OAuth2Token` there though, my idea was to store original object representing token, like `Jwt`, `String`, or class representing Paseo token in the future in the metadata. This has no immidate value but can be used later by Token Introspection endpoint. With that in mind, what about storing token like `.metadata(OAuth2TokenMetadata.TOKEN, jwt)`?

Reply:
After having a chat with Joe, we decided to keep the original `jwt` token in `TOKEN` attribute.